### DISCUSSION TITLE
Why do brew commands e.g install, cleanup not autocomplete in bash when i press tab

### DISCUSSION THREAD

**Discussion by ghost** [Homebrew/discussions#1168] [Linux] [answered]
`brew <TAB>` doesn't output anything

**Comment by carlocab** [comment, marked as answer]
Try https://docs.brew.sh/Shell-Completion#configuring-completions-in-bash.

**Reply by carlocab** [reply]
Not that I know of. You could switch to Zsh; preferably the Homebrew one, as that should work with completions for `brew` out of the box.

What does `brew config` say?

**Reply by carlocab** [reply]
Hmm. Actually, it seems the docs are outdated. Try
```sh
source $(brew --repo)/completions/bash/brew
```

**Reply by Rylan12** [reply]
You shouldn't need to do that. The instructions work fine for me.

`brew update` should automatically link the `$(brew --repo)/completions/bash/brew` file to `$(brew --prefix)/etc/bash_completion.d/bash` which should be loaded by the instructions in the docs.

Can you check to see whether you have the `$(brew --prefix)/etc/bash_completion.d/bash` file properly linked? Also, worth checking if you have `$(brew --prefix)/etc/profile.d/bash_completion.sh` as that would take precedence. I don't think Homebrew writes to the latter file, though.

Edit: I mistyped. The file is linked to `$(brew --prefix)/etc/bash_completion.d/brew` (instead of `.../bash`)

**Reply by carlocab** [reply]
Ah, this might be the problem. On my system, there is no `$(brew --prefix)/etc/bash_completion.d/bash`. There is, however, a `$(brew --prefix)/etc/bash_completion.d/brew`.

**Reply by Rylan12** [reply]
Sorry, I mistyped. I meant the latter (`brew`). I've edited the comment to fix that.

@brewfixvim what is the output of `ls -l $(brew --prefix)/etc/bash_completion.d`?

Can you also paste the output of `brew update`?

**Reply by Rylan12** [reply]
Are you pressing tab once or twice?

For bash completions you need to press tab twice:

```console
$ brew<TAB><TAB>
Display all 118 possibilities? (y or n)
```

**Reply by Rylan12** [reply]
What did you change?

Not sure why that would be the case but I guess it depends on your `~/.bash_profile` files. I can take a look but I'm definitely not an expert so might not be able to help much.

**Reply by ghost** [reply]
I've solved the issue. The problem was that bash wasn't sourcing its own completions.
All in all, if you'd like all kinds of bash completions use this script - https://docs.brew.sh/Shell-Completion#configuring-completions-in-bash together with this one;
```
if ! shopt -oq posix; then
  if [ -f /usr/share/bash-completion/bash_completion ]; then
    . /usr/share/bash-completion/bash_completion
  elif [ -f /etc/bash_completion ]; then
    . /etc/bash_completion
  fi
fi
```
